Add low product stock indicator

joshuatf opened the pull request:
Fixes #1827 

* Adds a low stock option to the products REST API
* Adds totalCount to items in wc-api
* Adds the unread indicator to the stock panel
* Extracts previous unread indicators to separate file

### Screenshots
<img width="402" alt="Screen Shot 2019-03-21 at 3 14 39 PM" src="https://user-images.githubusercontent.com/10561050/54737672-16e54500-4bec-11e9-8fc4-32c5e2c54b18.png">

### Detailed test instructions:

#### Products API
1. Make a request to the products controller with `low_in_stock` param set to a true value `wp-json/wc/v4/products/?low_in_stock=1`
2. Check that products returned are in fact low in stock.
3. Make sure the products returned respect the WC low stock threshold.
4. Set a low stock threshold unique to a product.
5. Make sure that the product only shows up under low stock when its value is under its specific threshold.

#### Stock unread indicator
1. Add at least some products that are considered low in stock.
2. Check that the unread indicator for the stock status is present.
3. Adjust stock settings so that none are considered low.
4.  Check that the unread indicator is no longer present.
5. Check that no regressions have occurred with the other indicators (due to minor refactor).

### Thoughts
* My original plan was to use an option to mark when a product became low stock, but this would become tricky with many out of stock products and would require extra queries on each order placement and product updates.  Plus this groundwork would have been needed anyway for querying these products once we actually implement the stock panel.
* The current solution will probably benefit from caching.  I initially set this to use a transient in https://github.com/woocommerce/woocommerce-admin/commit/19c8228b4e87eceb905abb1b5429b2cd71dd2f11, but we lose the ability to poll this data with freshdata.  We might want to follow this one up with some caching since this is a relatively strenuous query.

joshuatf committed:
Move product low stock count to controller param instead of data store

joshuatf committed:
Add totalCount to items API

joshuatf committed:
Extract unread indicators to separate file and add unread stock

jeffstieler commented on includes/api/class-wc-admin-rest-products-controller.php:
```diff
@@ -89,50 +98,67 @@ protected function prepare_objects_query( $request ) {
+				lis_postmeta3.meta_key = '_low_stock_amount' AND lis_postmeta.meta_value IS NOT NULL
+				AND lis_postmeta3.meta_key = '_low_stock_amount' AND CAST(lis_postmeta.meta_value AS SIGNED) <= CAST(lis_postmeta3.meta_value AS SIGNED)
+				OR lis_postmeta3.meta_key = '_low_stock_amount' AND lis_postmeta.meta_value IS NULL
+				AND lis_postmeta.meta_key = '_stock' AND CAST(lis_postmeta.meta_value AS SIGNED) <= '{$low_stock_amount}'
```
Looks like `$low_stock_amount` didn't make the jump from https://github.com/woocommerce/woocommerce-admin/pull/1866/commits/19c8228b4e87eceb905abb1b5429b2cd71dd2f11#diff-04ebb0cb4b33c92c166ab779a502d791R530.

Also, could we delete the first two commits?

joshuatf replied:
Whoops- thanks!

>Also, could we delete the first two commits?

Yep.  Wanted to make sure this was the right direction before dropping those.

jeffstieler commented on includes/api/class-wc-admin-rest-products-controller.php:
```diff
@@ -145,14 +171,14 @@ public static function add_wp_query_product_search_join( $join, $wp_query ) {
-		$search = trim( $wp_query->get( 'search' ) );
-		if ( $search && empty( $groupby ) ) {
+		$search       = trim( $wp_query->get( 'search' ) );
+		$low_in_stock = trim( $wp_query->get( 'low_in_stock' ) );
```
Why `trim()` here? If `prepare_objects_query()` handled coercion, would this be needed?

joshuatf replied:
You're right, not necessary.

jeffstieler commented on includes/api/class-wc-admin-rest-products-controller.php:
```diff
@@ -78,6 +84,9 @@ protected function prepare_objects_query( $request ) {
 			$args['search'] = $request['search'];
```
Should there be a `trim()` here? _(related to a comment below)_

joshuatf replied:
It's already trimmed elsewhere, but we can trim here instead and remove the other trims to make this more manageable.

jeffstieler commented on includes/api/class-wc-admin-rest-products-controller.php:
```diff
@@ -78,6 +84,9 @@ protected function prepare_objects_query( $request ) {
 			unset( $args['s'] );
 		}
+		if ( ! empty( $request['low_in_stock'] ) ) {
+			$args['low_in_stock'] = $request['low_in_stock'];
```
Should this be coerced to a boolean?

joshuatf committed:
Trim initial search variable in wp_query

joshuatf commented:
@jeffstieler Thanks for the review!  This one has been rebased and is ready for another look.

jeffstieler commented on includes/api/class-wc-admin-rest-products-controller.php:
```diff
@@ -89,50 +98,68 @@ protected function prepare_objects_query( $request ) {
 		}
 
+		if ( $wp_query->get( 'low_in_stock' ) ) {
+			$low_stock_amount = get_option( 'woocommerce_notify_low_stock_amount', 2 );
```
We need to either SQL escape this or parse it as an integer (`get_option` is filterable).

joshuatf replied:
Updated in https://github.com/woocommerce/woocommerce-admin/pull/1866/commits/e466d4029fda1ef2528af988060c2391c3706cd4

joshuatf committed:
Parse low stock option as an absolute integer

jeffstieler commented on includes/api/class-wc-admin-rest-products-controller.php:
```diff
@@ -127,7 +127,7 @@ public static function add_wp_query_filter( $where, $wp_query ) {
 
 		if ( $wp_query->get( 'low_in_stock' ) ) {
-			$low_stock_amount = get_option( 'woocommerce_notify_low_stock_amount', 2 );
+			$low_stock_amount = absint( max( get_option( 'woocommerce_notify_low_stock_amount', 2 ) ) );
```
`max()` is unnecessary here and will throw a warning if the first argument isn't an array.

```
php > var_dump( max( 'a' ) );
Warning: max(): When only one parameter is given, it must be an array
```

joshuatf replied:
@jeffstieler Changed this back to not use `max()`, but do you know why they do this throughout core?  I assumed there was some plugin that used an array as most instances are written as:

```php
absint( max( get_option( 'woocommerce_notify_low_stock_amount', 2 ) ) );
```

jeffstieler replied:
@joshuatf the core uses of `max()` provide a second argument: https://github.com/woocommerce/woocommerce/blob/f9cdab66eb1f61f6f26001838199aa1a5b8ba01f/includes/admin/reports/class-wc-report-low-in-stock.php#L39

From https://www.php.net/max:
> If the first and only parameter is an array, max() returns the highest value in that array. If at least two parameters are provided, max() returns the biggest of these values.

joshuatf replied:
🤦‍♂️ I kept reading this as the second parameter that acts as the default value in `get_option()`.  Going to swap this for a max with `1` since this seems to be the most commonly used in core for low stock reporting.

jeffstieler commented:
Thanks for sticking through this long review @joshuatf.

I have some questions about the query being run for low stock products:

_(I've formatted this by hand, hopefully correctly, and added comments)_

```sql
SELECT SQL_CALC_FOUND_ROWS wp_posts.ID
FROM wp_posts
INNER JOIN wp_postmeta AS lis_postmeta
	ON wp_posts.ID = lis_postmeta.post_id
INNER JOIN wp_postmeta AS lis_postmeta2
	ON wp_posts.ID = lis_postmeta2.post_id
INNER JOIN wp_postmeta AS lis_postmeta3
	ON wp_posts.ID = lis_postmeta3.post_id
WHERE 1=1
AND wp_posts.post_type = 'product'
AND ((wp_posts.post_status = 'publish'))
AND lis_postmeta2.meta_key = '_manage_stock'
AND lis_postmeta2.meta_value = 'yes'
AND lis_postmeta.meta_key = '_stock'
AND lis_postmeta.meta_value IS NOT NULL
AND (
	lis_postmeta3.meta_key = '_low_stock_amount'
	AND lis_postmeta.meta_value IS NOT NULL # redundant condition outside of parens
	AND lis_postmeta3.meta_key = '_low_stock_amount' # redundant with 2 lines above
	AND CAST(lis_postmeta.meta_value AS SIGNED) <= CAST(lis_postmeta3.meta_value AS SIGNED)
	OR
	lis_postmeta3.meta_key = '_low_stock_amount'
	AND lis_postmeta.meta_value IS NULL # contradictory with condition outside of parens
	AND lis_postmeta.meta_key = '_stock' # redundant condition outside of parens
	AND CAST(lis_postmeta.meta_value AS SIGNED) <= '2' # casting null and comparing?
)
GROUP BY wp_posts.ID
ORDER BY wp_posts.post_date DESC, wp_posts.ID DESC
LIMIT 0, 1
```

I'm noticing some redundant/contradictory conditions in here, mostly around `lis_postmeta.meta_value` being null or not. Can you double check this?

Also, `$low_stock_amount` is a string in the query, not sure if that's intentional.

joshuatf committed:
Fix low stock meta comparison with product specific low stock amount

joshuatf commented:
@jeffstieler Thanks for the review! 
* The post meta null comparisons were meant to be against `lis_postmeta3`.  Fixed those up.
* Also removed/extracted some of those redundant conditions.

```sql
AND lis_postmeta2.meta_key = '_manage_stock'
AND lis_postmeta2.meta_value = 'yes'
AND lis_postmeta.meta_key = '_stock'
AND lis_postmeta.meta_value IS NOT NULL
AND lis_postmeta3.meta_key = '_low_stock_amount'
AND (
	lis_postmeta3.meta_value IS NOT NULL
	AND CAST(lis_postmeta.meta_value AS SIGNED) <= CAST(lis_postmeta3.meta_value AS SIGNED)
	OR lis_postmeta3.meta_value IS NULL
	AND CAST(lis_postmeta.meta_value AS SIGNED) <= {$low_stock_amount}
)
```

joshuatf committed:
Make group by query function static

joshuatf committed:
Check for empty strings and null values for product low_stock_amount

jeffstieler reviewed (approved):
Everything is looking good! :shipit:

jeffstieler commented:
I'm going to resolve the conflict and merge this - it will be really helpful for hooking up the stock panel.

joshuatf commented:
Thanks for reviewing and fixing those conflicts @jeffstieler!
